torch/csrc/autograd/record_function.cpp

```diff
@@ -4,20 +4,23 @@
 namespace torch { namespace autograd { namespace profiler {
 
 namespace {
-bool has_callbacks = false;
+RecordFunctionState record_function_state = RecordFunctionState::NO_CALLBACKS;
 std::vector<RecordFunctionCallback> start_callbacks;
 std::vector<RecordFunctionCallback> end_callbacks;
+std::vector<bool> needs_inputs_flags({false});
 thread_local RecordFunction* thread_local_func_ = nullptr;
 }
 
-void pushCallback(RecordFunctionCallback start, RecordFunctionCallback end) {
+void pushCallback(
+    RecordFunctionCallback start, RecordFunctionCallback end, bool needs_inputs) {
   start_callbacks.push_back(start);
   end_callbacks.push_back(end);
-  has_callbacks = true;
-}
-
-void pushCallback(RecordFunctionCallback start) {
-  pushCallback(start, [](const RecordFunction&){});
+  needs_inputs_flags.push_back(needs_inputs_flags.back() || needs_inputs);
```

Comment:
It looks like the only valid values this vector can take is a sequence of `false` followed by a sequence of `true`. Why can't you replace it with a counter? In here, it would look like (assuming you'd rename it to `callbacks_need_input`:
```cpp
if (callbacks_need_input || needs_inputs) {
  ++callbacks_need_input;
}
```

Reply:
If you pop a callback, what would you do with this counter? decrease it? or not?

Alternative solution that I see is to remember the first time we saw a callback that needs an input is pushed on the stack